File: 72.编辑距离.java
/*
 * @lc app=leetcode.cn id=72 lang=java
 *
 * [72] 编辑距离
 */

// @lc code=start
class Solution {
    public int minDistance(String word1, String word2) {
        int m = word1.length(), n = word2.length();
        int[][] dp = new int[m + 1][n + 1];

        // base cases: edit distance from empty string to target string s[:i] or s[:j]
        for (int i = 0; i <= m; ++i) {
            dp[i][0] = i;
        }
        for (int j = 0; j <= n; ++j) {
            dp[0][j] = j;
        }

        for (int i = 1; i <= m; ++i) {
            for (int j = 1; j <= n; ++j) {
                if (word1.charAt(i - 1) == word2.charAt(j - 1)) {
                    // skip the current identical characters for both words
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    dp[i][j] = min(
                        dp[i - 1][j] + 1,       // delete one character from word1
                        dp[i][j - 1] + 1,       // add one character to word2
                        dp[i - 1][j - 1] + 1    // replace one character
                    );
                }
            }
        }
        return dp[m][n];
    }

    private int min(int opA, int opB, int opC) {
        return Math.min(opA, Math.min(opB, opC));
    }
}
// @lc code=end

File: 72.编辑距离.py
#
# @lc app=leetcode.cn id=72 lang=python3
#
# [72] 编辑距离
#

# @lc code=start
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n1, n2 = len(word1), len(word2)
        dp = [[0]*(n2+1) for _ in range(n1+1)]
        # 第一行、第一列
        for i in range(1, n1 + 1):
            dp[i][0] = dp[i-1][0] + 1
        for j in range(1, n2 + 1):
            dp[0][j] = dp[0][j-1] + 1
        
        # 一般情况
        for i in range(1, n1+1):
            for j in range(1, n2+1):
                if word1[i-1] == word2[j-1]:
                    dp[i][j] = dp[i-1][j-1]
                else:
                    dp[i][j] = min(dp[i-1][j-1], dp[i-1][j], dp[i][j-1]) + 1
        return dp[-1][-1]


# @lc code=end
print(Solution().minDistance('intention', 'execution'))
